src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexNodes.cs

```diff
@@ -0,0 +1,1643 @@
+    internal abstract class RegexGroupingNode : RegexPrimaryExpressionNode
+    {
+        protected RegexGroupingNode(RegexKind kind, RegexToken openParenToken, RegexToken closeParenToken)
+            : base(kind)
+        {
+            Debug.Assert(openParenToken.Kind == RegexKind.OpenParenToken);
+            Debug.Assert(closeParenToken.Kind == RegexKind.CloseParenToken);
+            OpenParenToken = openParenToken;
+            CloseParenToken = closeParenToken;
+        }
+
+        public RegexToken OpenParenToken { get; }
+        public RegexToken CloseParenToken { get; }
+    }
+
+    /// <summary>
+    /// The ```(...)``` node you get when the group does not start with ```(?```
+    /// </summary>
+    internal class RegexSimpleGroupingNode : RegexGroupingNode
+    {
+        public RegexSimpleGroupingNode(RegexToken openParenToken, RegexExpressionNode expression, RegexToken closeParenToken)
+            : base(RegexKind.SimpleGrouping, openParenToken, closeParenToken)
+        {
+            Debug.Assert(expression != null);
+            Expression = expression;
+        }
+
+        public RegexExpressionNode Expression { get; }
+
+        internal override int ChildCount => 3;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+            case 0: return OpenParenToken;
+            case 1: return Expression;
+            case 2: return CloseParenToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// Base type of all ```(?...)``` groupings.
+    /// </summary>
+    internal abstract class RegexQuestionGroupingNode : RegexGroupingNode
+    {
+        protected RegexQuestionGroupingNode(RegexKind kind, RegexToken openParenToken, RegexToken questionToken, RegexToken closeParenToken)
+            : base(kind, openParenToken, closeParenToken)
+        {
+            Debug.Assert(questionToken.Kind == RegexKind.QuestionToken);
+            QuestionToken = questionToken;
+        }
+
+        public RegexToken QuestionToken { get; }
+    }
+
+    /// <summary>
+    /// Base type of ```(?inmsx)``` or ```(?inmsx:...)``` nodes.
+    /// </summary>
+    internal abstract class RegexOptionsGroupingNode : RegexQuestionGroupingNode
+    {
+        protected RegexOptionsGroupingNode(RegexKind kind, RegexToken openParenToken, RegexToken questionToken, RegexToken optionsToken, RegexToken closeParenToken)
+            : base(kind, openParenToken, questionToken, closeParenToken)
+        {
+            OptionsToken = optionsToken;
+        }
+
+        public RegexToken OptionsToken { get; }
+    }
+
+    /// <summary>
+    /// ```(?inmsx)``` node.  Changes options in a sequence for all subsequence nodes.
+    /// </summary>
+    internal class RegexSimpleOptionsGroupingNode : RegexOptionsGroupingNode
```

Comment:
`sealed`? (this applied to other nodes in the file that aren't `abstract`)

Reply:
will do!

Reply:
#resolved